### PR TITLE
Fix Usage failed to get pid

### PR DESCRIPTION
**Pull request by div8cn**
## Description

cloudstack-usage use System.getProperty("pid") to get the process PID
System.getProperty cannot get the Pid, which will cause the service to stay in the config stage. Cannot continue to start, and there will be no log output.

This PR uses ManagementFactory instead of System.getProperty to get Pid, and throws an exception when Pid cannot be obtained

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

**Comment by rhtyd** [comment]
@div8cn where did you reproduce the issue, OS/distro, CloudStack version etc? Thanks

**Comment by div8cn** [comment]
I encountered this problem when building cloudstack-usage into docker and started using supervisord。

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: usage/src/main/java/com/cloud/usage/UsageManagerImpl.java [hunk 1]
```diff
@@ -275,7 +277,14 @@ public boolean configure(String name, Map<String, Object> params) throws Configu
...
             throw new ConfigurationException("Unhandled exception configuring UsageManager " + e.toString());
         }
-        _pid = Integer.parseInt(System.getProperty("pid"));
+
+        try {
+            String processName = ManagementFactory.getRuntimeMXBean().getName();
+            _pid = Integer.parseInt(processName.split("@")[0]);
+        } catch (Exception e) {
+            s_logger.error("Unable to get process pid ", e);
+            throw new ConfigurationException("Unable to get process pid " + e.toString());
```
Comment (ravening) [inline usage/src/main/java/com/cloud/usage/UsageManagerImpl.java:286]: Should we add a more meaningful message here so its more clear?

**Comment by rhtyd** [comment]
@div8cn can you check if in the docker container the pid parameter was passed? Depending on how the service runs perhaps this was not passed, look at the cloudstack-usage systemd service for example.

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: usage/src/main/java/com/cloud/usage/UsageManagerImpl.java [hunk 2]
```diff
@@ -275,7 +277,14 @@ public boolean configure(String name, Map<String, Object> params) throws Configu
...
+        }
```
Comment (DaanHoogland) [inline usage/src/main/java/com/cloud/usage/UsageManagerImpl.java:287]: 1. twice the same message should be in a string constand
1. rethrown should not abandon the stack trace for the original problem
```suggestion
        String processName;
        try {
            processName = ManagementFactory.getRuntimeMXBean().getName();
            _pid = Integer.parseInt(processName.split("@")[0]);
        } catch (Exception e) {
            String msg = String.format("Unable to get process Id for %s!", processName);
            s_logger.error(msg);
            throw new ConfigurationException(msg, e);
        }
```
or
```suggestion
        String processName;
        try {
            processName = ManagementFactory.getRuntimeMXBean().getName();
            _pid = Integer.parseInt(processName.split("@")[0]);
        } catch (Exception e) {
            String msg = String.format("Unable to get process Id for %s!", processName);
            s_logger.debug(msg , e);
            throw new ConfigurationException(msg, e);
        }
```

File: usage/src/main/java/com/cloud/usage/UsageManagerImpl.java (hunk 1 above)
Comment (DaanHoogland) [reply usage/src/main/java/com/cloud/usage/UsageManagerImpl.java]: I'm all for clarity @ravening . what do you suggest?

File: usage/src/main/java/com/cloud/usage/UsageManagerImpl.java [hunk 3]
```diff
@@ -275,7 +277,16 @@ public boolean configure(String name, Map<String, Object> params) throws Configu
...
-        _pid = Integer.parseInt(System.getProperty("pid"));
+
+        String processName = null;
+        try {
+            processName = ManagementFactory.getRuntimeMXBean().getName();
+            _pid = Integer.parseInt(processName.split("@")[0]);
+        } catch (Exception e) {
+            String msg = String.format("Unable to get process Id for %s!", processName);
+            s_logger.debug(msg , e);
+            throw new ConfigurationException(msg + " " + e.toString());
```
Comment (GabrielBrascher) [inline usage/src/main/java/com/cloud/usage/UsageManagerImpl.java:288]: @DaanHoogland @div8cn What do you think of using `e.printStackTrace()` to add the stack on the log/ConfigurationException?

With `toString` (or simply s_logger.debug(msg , e)) it logs only the exception name and the error message (if there is any message). On the other hand, `printStackTrace()` presents the whole stack trace of an exception, which is very helpful for debugging.

Comment (DaanHoogland) [reply usage/src/main/java/com/cloud/usage/UsageManagerImpl.java:288]: `logger.debug(msg, e)` should log the entire stacktrace @GabrielBrascher . and it gives the user to filter or direct the output to a certain stream/file. `e.printStacktrace()` only logs to stdout (or stderr, i forgot). in short i think not too much of it. I agree toString does not seem the best, but we have alreadyt output the stacktrace by then. A change from `toString()` to `getLocalizedMessage()` maybe, @div8cn? Of course we can add the entire exception again i.e. `throw new ConfigurationException(msg, e)`, that's fine as well.

**Comment by rhtyd** [comment]
@blueorangutan package

File: usage/src/main/java/com/cloud/usage/UsageManagerImpl.java [hunk 4]
```diff
@@ -275,7 +277,16 @@ public boolean configure(String name, Map<String, Object> params) throws Configu
             s_logger.error("Unhandled exception configuring UsageManger", e);
             throw new ConfigurationException("Unhandled exception configuring UsageManager " + e.toString());
         }
-        _pid = Integer.parseInt(System.getProperty("pid"));
+
+        String processName = null;
+        try {
+            processName = ManagementFactory.getRuntimeMXBean().getName();
```
Comment (rhtyd) [inline usage/src/main/java/com/cloud/usage/UsageManagerImpl.java:283]: @div8cn master is on java 11 now, can you use `ProcessHandle.current().pid()`? I did a quick test:
```
> jshell 
|  Welcome to JShell -- Version 11.0.7
|  For an introduction type: /help intro

jshell> ProcessHandle.current().pid()
$1 ==> 20380

```

**Comment by div8cn** [comment]
@rhtyd  Thanks, I've updated it to ProcessHandle.current ().pid().

File: usage/src/main/java/com/cloud/usage/UsageManagerImpl.java [hunk 5]
```diff
@@ -280,8 +278,7 @@ public boolean configure(String name, Map<String, Object> params) throws Configu
 
         String processName = null;
         try {
-            processName = ManagementFactory.getRuntimeMXBean().getName();
-            _pid = Integer.parseInt(processName.split("@")[0]);
+            _pid = (int) ProcessHandle.current().pid();
         } catch (Exception e) {
             String msg = String.format("Unable to get process Id for %s!", processName);
```
Comment (weizhouapache) [inline usage/src/main/java/com/cloud/usage/UsageManagerImpl.java:283]: processName is always null .

File: usage/src/main/java/com/cloud/usage/UsageManagerImpl.java [hunk 6]
```diff
@@ -275,7 +275,15 @@ public boolean configure(String name, Map<String, Object> params) throws Configu
             s_logger.error("Unhandled exception configuring UsageManger", e);
             throw new ConfigurationException("Unhandled exception configuring UsageManager " + e.toString());
         }
-        _pid = Integer.parseInt(System.getProperty("pid"));
+
+        String processName = null;
```
Comment (rhtyd) [inline usage/src/main/java/com/cloud/usage/UsageManagerImpl.java:279]: processName is not used and can be removed @div8cn

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Review by rhtyd — APPROVED** [review]

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
though the ha error is a bit offsetting I see not how any of these errors can be related, @rhtyd agree?

**Review by rhtyd — APPROVED** [review]

**Comment by rhtyd** [comment]
Yup LGTM

**Pull request merged**